modules/querier/external/token_provider.go

```diff
@@ -0,0 +1,73 @@
+package external
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"go.uber.org/multierr"
+	"golang.org/x/oauth2"
+	"google.golang.org/api/idtoken"
+)
+
+// Caches an oauth2.TokenSource to enable efficient auth on each of our external
+// endpoints.
+type cachedTokenProvider struct {
+	tokenSources map[string]oauth2.TokenSource
+}
+
+func newTokenProvider(
+	ctx context.Context,
+	endpoints []string,
+	getTokenSource func(ctx context.Context, endpoint string) (oauth2.TokenSource, error),
+) (*cachedTokenProvider, error) {
+	sources := make(map[string]oauth2.TokenSource, len(endpoints))
+
+	var mtx sync.Mutex
+	var wg sync.WaitGroup
+
+	var tsErr error
+	for _, endpoint := range endpoints {
+		wg.Add(1)
+		go func(ep string) {
+			defer wg.Done()
+			ts, err := getTokenSource(ctx, ep)
+
+			mtx.Lock()
+			defer mtx.Unlock()
+
+			if err != nil {
+				tsErr = multierr.Combine(tsErr, err)
+			}
+			sources[ep] = ts
+		}(endpoint)
+	}
+	wg.Wait()
```

Comment:
Note that this code is only invoked when the app starts up initially, so performance isn't a huge concern, so I'm happy to remove the goroutine + WaitGroup if we want to simplify. But I figured concurrent execution was worthwhile since `getTokenSource` can be a blocking call.

For example, when the `idtoken.NewTokenSource` function (used to auth into Google's api) is passed as `getTokenSource` (as done in the `newGoogleProvider` function below), it'll [perform 2 legged auth to retrieve the token](https://github.com/googleapis/google-api-go-client/blob/ed3a635a8d60c849d27763cc619105088abb1e06/idtoken/idtoken.go#L137-L138) as well as [read from a credentials file on disk](https://github.com/googleapis/google-api-go-client/blob/ed3a635a8d60c849d27763cc619105088abb1e06/idtoken/idtoken.go#L102).

Reply:
no, i think this is a good call. this kind of structure is common and does not require much mental overhead to parse.